kbase-extension/static/kbase/js/util/icon.js

```diff
@@ -1,130 +1,181 @@
-        };
-        if (isCustomIcon(icon)) {
-            // add custom icons (more than 1 will look weird, though)
-            _.each(icon, function(cls) { addLogoFunc('icon', $logo, cls); });
-        } else {
-            // add stack of font-awesome icons
-            _.each(icon, function(cls) { addLogoFunc('fa', $logo, cls); });
-        }
+    /**
+     * @method _makeIcon
+     * @private
+     * Creates an HTML string that renders an icon
+     * @param {string} iconType  the type of icon to produce,
+     *                              e.g. 'app', 'app-toolbar', 'data', 'generic', etc.
+     * @param {object} iconClass  the class of the icon, e.g. 'fa-cube'
+     * @param {string} shape      shape for the icon background;
+     *                              typically 'square' or 'circle'
+     * @param {string} color      colour for the icon background (hex or RGB)
+     * @param {bool}   stacked    whether or not the icon should be stacked
+     */
+
+    function _makeIcon(iconType, iconClass, shape = 'square', color = null, stacked = false) {
+
+        const style = color ? {color: color} : {};
+        const layers = stacked ? _makeStackLayers(shape, color) : [];
+        iconClass = iconClass.replace('icon ', '');
+
+        return span({
+            class: `${cssBaseName}__container--${iconType} fa-stack`,
+        }, layers.concat(
+            span({
+                class: `${cssBaseName}__icon_background--${iconType} fa fa-${shape} fa-stack-2x`,
+                style: style,
+            }),
+            span({
+                class: `${cssBaseName}__icon--${iconType} fa fa-inverse fa-stack-1x ${iconClass}`,
+            })
+        ));
+    }
+
+    /**
+     * @method
+     * @public
+     * Builds a data icon for a given data type
+     * @param {string} type The shortened type of object to fetch the icon for. E.g. "Genome" not
+     * "KBaseGenomes.Genome-2.1"
+     * @param {boolean} stacked If true, creates a stacked effect of multiple 'icons'.
+     */
+    function makeDataIcon(type, stacked) {
+        const iconType = stacked ? 'data-stack' : 'data',
+            color = _.has(iconSpec.color_mapping, type)
+                ? iconSpec.color_mapping[type]
+                : logoColorLookup(type),
+            icon = _.has(iconSpec.data, type)
+                ? iconSpec.data[type]
+                : iconSpec.data.DEFAULT;
+        return _makeIcon(iconType, icon[0], 'circle', color, stacked);
+    }
+
+    /**
+     * @method
+     * @public
+     * Builds a data icon in the given $logo div.
+     * @param {object} $logo jQuery element in which to insert an icon.
+     * @param {string} type The shortened type of object to fetch the icon for. E.g. "Genome" not
+     * "KBaseGenomes.Genome-2.1"
+     * @param {boolean} stacked If true, creates a stacked effect of multiple 'icons'.
+     * @param {int} indent -- deprecated
+     */
+    function buildDataIcon($logo, type, stacked, indent) {
```

Comment:
Left this in for compatibility with existing code (to reduce the size of the PR!). I intend to remove it in the next PR as it can be replaced by `makeDataIcon`.

Reply:
Great!